setup stats command

fahedouch opened the pull request:
Signed-off-by: fahed dorgaa <fahed.dorgaa@gmail.com>

AkihiroSuda commented on stats.go:
```diff
@@ -0,0 +1,386 @@
+package main
+
+import (
+	//"encoding/json"
```
Could you clean up comment lines

AkihiroSuda commented on stats.go:
```diff
@@ -0,0 +1,386 @@
+/*
+   Copyright (C) nerdctl authors.
```
This file seems derivative of containerd, right?

fahedouch replied:
not 100%, several logic I introduced them myself

fahedouch replied:
You want to add containerd Copyright ?

AkihiroSuda replied:
Yes

AkihiroSuda commented on formatter_stats.go:
```diff
@@ -0,0 +1,225 @@
+	cpuPercHeader   = "CPU %"
+	netIOHeader     = "NET I/O"
+	blockIOHeader   = "BLOCK I/O"
+	memPercHeader   = "MEM %"             // Used only on Linux
```
I'd prefer not to add Windows-specific code at the moment (until we support `nerdctl run` on Windows)

AkihiroSuda commented on formatter_stats.go:
```diff
@@ -0,0 +1,225 @@
+package main
```
Can we have this under a separate pkg like `pkg/stats`?

AkihiroSuda commented on go.mod:
```diff
@@ -18,8 +19,10 @@ require (
+	github.com/jaguilar/vt100 v0.0.0-20201024211400-81de19cb81a4 // indirect
 	github.com/lib/pq v1.9.0 // indirect
 	github.com/mattn/go-isatty v0.0.12
+	github.com/moby/buildkit v0.8.1 // indirect
```
Why?

AkihiroSuda commented on go.mod:
```diff
@@ -32,15 +35,27 @@ require (
 	github.com/stretchr/testify v1.7.0 // indirect
 	github.com/theupdateframework/notary v0.7.0 // indirect
+	github.com/tonistiigi/go-rosetta v0.0.0-20201102221648-9ba854641817 // indirect
+	github.com/urfave/cli v1.22.2
```
Why?

AkihiroSuda commented on pkg/stats/formatter_stats.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+// SetError sets container statistics error
+func (cs *Stats) SetError(err error) {
```
Unused function?

AkihiroSuda commented on pkg/stats/formatter_stats.go:
```diff
@@ -0,0 +1,214 @@
+
+// SetErrorAndReset zeroes all the container statistics and store the error.
+// It is used when receiving time out error during statistics collecting to reduce lock overhead
+func (cs *Stats) SetErrorAndReset(err error) {
```
Unused function?

AkihiroSuda commented on stats.go:
```diff
@@ -0,0 +1,423 @@
+		var err error
+
+		if runtime.GOOS == "windows" {
+			fmt.Println("nerdctl do not yet support windows")
```
This is not specific to `stats`, and I don't even think it compiles on Windows, so this Println is not needed

AkihiroSuda commented on stats.go:
```diff
@@ -0,0 +1,423 @@
+			case *v2.Metrics:
+				data2 = v
+			default:
+				//return errors.New("cannot convert metric data to cgroups.Metrics or windows.Statistics")
```
Please clean up comments

AkihiroSuda commented on stats.go:
```diff
@@ -0,0 +1,423 @@
+				pidsStatsCurrent = data.Pids.Current
+				netRx, netTx = calculateNetwork(data)
+			} else if data2 != nil {
+				//previousCPU = data2.CPU.UsageUsec
```
Is this WIP?

AkihiroSuda commented on stats.go:
```diff
@@ -0,0 +1,423 @@
+
+		format := optionsStats.format
+		if len(format) == 0 {
+			/*if len(dockerCli.ConfigFile().StatsFormat) > 0 {
```
Please clean up comments

AkihiroSuda commented on pkg/stats/formatter_stats.go:
```diff
@@ -0,0 +1,214 @@
+	if c.s.IsInvalid {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%d", c.s.PidsCurrent)
```
Should take clicontext.App.Writer

AkihiroSuda commented:
Thanks for working on this, but needs cleaning up the code, squashing the commits, and rebasing with the master.

AkihiroSuda commented:
@fahedouch Are you still planning to work on this?

fahedouch commented:
@AkihiroSuda yes, if you see if it's a priority I can prioritize it over other tasks

nehaldattani commented:
This enhancement is pending for quite a while.   Would this PR be accepted anytime soon ?

fahedouch commented:
I did the necessary except the stats for the `windows container`; some statistics are missing, I am trying to find out how to do the formulas to have the right stats for windows containers

AkihiroSuda commented on cmd/nerdctl/stats.go:
```diff
@@ -0,0 +1,427 @@
+	"text/template"
+	"time"
+
+	wstats "github.com/Microsoft/hcsshim/cmd/containerd-shim-runhcs-v1/stats"
```
Can we split this to `*_windows.go` and `*_linux.go`?

AkihiroSuda commented:
Thanks for working on this, but this doesn't seem to show containers that are created after starting `nerdctl stats`.

Windows metrics can be added later. (We don't even have `nerdctl run` for Windows yet)

AkihiroSuda commented:
Also, I see strange CPU stat like `11496300.00%`.
(Only tested cgroup v2, rootless)

dcantah commented on cmd/nerdctl/stats.go:
```diff
@@ -0,0 +1,427 @@
+		SilenceErrors:     true,
+	}
+
+	addStastFlags(statsCommand)
```
stats*

fahedouch replied:
fixed

dcantah commented on cmd/nerdctl/stats.go:
```diff
@@ -0,0 +1,427 @@
+	return statsCommand
+}
+
+func addStastFlags(cmd *cobra.Command) {
```
stats*

fahedouch replied:
fixed

dcantah commented on cmd/nerdctl/stats.go:
```diff
@@ -0,0 +1,427 @@
+	// NOTE: rootless container does not rely on cgroupv1.
+	// more details about possible ways to resolve this concern: #223
+	if rootlessutil.IsRootless() && infoutil.CgroupsVersion() == "1" {
+		return fmt.Errorf("stats requires cgroup v2 for rootless containers, see https://rootlesscontaine.rs/getting-started/common/cgroup2/")
```
Nit: errors.New if no formatting needed

fahedouch replied:
fixed

dcantah commented on pkg/statsutil/stats.go:
```diff
@@ -0,0 +1,380 @@
+
+// Stats represents an entity to store containers statistics synchronously
+type Stats struct {
+	mutex sync.Mutex
```
Would an RWMutex be preferred here? Seems it would benefit the Get methods

fahedouch replied:
thanks, fixed

dcantah commented on pkg/statsutil/stats.go:
```diff
@@ -0,0 +1,380 @@
+	if limit != 0 {
+		return usedNo / limit * 100.0
+	}
+	return 5
```
Why 5?

fahedouch replied:
fixed

dcantah commented on cmd/nerdctl/stats.go:
```diff
@@ -0,0 +1,427 @@
+				data = v
+			case *v2.Metrics:
+				data2 = v
+			case *wstats.WindowsContainerStatistics:
+				windowsStats = v
```
Part of me thinks it's odd to add Windows logic at the moment, as I don't believe this supports running containers Windows yet.

fahedouch commented:
> Thanks for working on this, but this doesn't seem to show containers that are created after starting `nerdctl stats`.
> 
> Windows metrics can be added later. (We don't even have `nerdctl run` for Windows yet)

 @AkihiroSuda  as I mentionned in this comment https://github.com/containerd/nerdctl/pull/73/files#diff-bab6d6355f721a47c2a974e88b96dde11fdeaa6a8d862ee3f7974558b00abc52R150 not yet supported. I will add it second PR if it is possible! 

By the way I fixed the Cpu stats for Cgroupv2

AkihiroSuda commented on cmd/nerdctl/stats_linux.go:
```diff
@@ -0,0 +1,413 @@
+	}
+
+	if showAll {
+		// TODO: watches for container creation and removal (dynamic stats)
```
Could you add this limitation to the `--help` text and `README.md`

AkihiroSuda replied:
And open an issue ticket

AkihiroSuda commented:
CI failing for Windows and FreeBSD

fahedouch commented:
@AkihiroSuda  I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.

AkihiroSuda commented on pkg/eventutil/eventutil.go:
```diff
@@ -0,0 +1,57 @@
+func (w *eventHandler) Watch(c <-chan *events.Envelope) {
+	for e := range c {
+		w.mu.Lock()
+		fmt.Println(e.Topic)
```
remove

AkihiroSuda commented:
> @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.

Thanks, but creation doesn't seem handled yet

AkihiroSuda commented on pkg/statsutil/stats_linux.go:
```diff
@@ -0,0 +1,160 @@
+	mem = calculateCgroup2MemUsage(metrics)
+	memLimit = float64(metrics.Memory.UsageLimit)
+	memPercent = calculateMemPercent(memLimit, mem)
+	pidsStatsCurrent = metrics.Pids.Current
```
NetworkTx/Rx missing?

fahedouch replied:
@AkihiroSuda  it seems that network is not supported by cgroupv2 https://pkg.go.dev/github.com/containerd/cgroups@v1.0.2/v2/stats#Metrics

AkihiroSuda replied:
But `docker stats` work with cgroup v2.
We can revisit this in another PR.

fahedouch replied:
@AkihiroSuda  I'm assuming that `docker` doesn't consume the cgroup api from containerd  https://github.com/containerd/cgroups. 
do you think it is a lack in the containerd cgroup api and that I have to open a ticket at  https://github.com/containerd/cgroups

AkihiroSuda replied:
docker scans the stats from netlink, not from cgroups

fahedouch replied:
@AkihiroSuda thank you for checking this I will fix it.

fahedouch commented:
> > @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.
> 
> Thanks, but creation doesn't seem handled yet

creation is handled only with `--all` flag,  have you tested with with `--all` flag ?

AkihiroSuda commented:
> > > @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.
> > 
> > 
> > Thanks, but creation doesn't seem handled yet
> 
> @AkihiroSuda creation is handled only with `--all` flag, have you tested with with `--all` flag ?

Can we watch creation events by default?
`docker stats` doesn't need `--all` to watch creation events.

fahedouch commented:
> > > > @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.
> > > 
> > > 
> > > Thanks, but creation doesn't seem handled yet
> > 
> > 
> > @AkihiroSuda creation is handled only with `--all` flag, have you tested with with `--all` flag ?
> 
> Can we watch creation events by default? `docker stats` doesn't need `--all` to watch creation events.

fixed @AkihiroSuda

AkihiroSuda commented on README.md:
```diff
@@ -931,6 +931,12 @@ Flags:
+Display a live stream of container(s) resource usage statistics.
+
+
+Usage: `nerdctl stats [flags]`
```
Next time please add `flags` too

AkihiroSuda reviewed (approved):
Thanks
